### ISSUE TITLE
Performance issues with `eval vm_compute` inside proof of a Derive command

### ISSUE THREAD

**Issue by DIJamner** [coq/coq#14455]
#### Description of the problem

The Ltac2 library function `Std.eval_vm` runs unexpectedly slowly compared to Ltac's `vm_compute` when its input performs certain computations that include the identifier of a Derive command. In the below example, the first 3 evaluations run instantly but the fourth takes notable time, increasing with its natural number argument.

```
Require Import List.
Import ListNotations.

Require Coq.derive.Derive.

Require Import Ltac2.Ltac2.
Set Default Proof Mode "Classic".

Inductive exp : Set := con : list exp -> exp.

Fixpoint run_list {A} (l : list A) : unit :=
  match l with
  | [] => tt
  | _::l' => run_list l'
  end.

Fixpoint run_nat (l : nat) : unit :=
  match l with
  | 0 => tt
  | S l' => run_nat l'
  end.

Fixpoint traverse_n_times (l : nat) n (e : exp) : exp :=
  match n with
  | 0 => con []
  | S n' =>
    let '(con s) := e in
    let () := run_list (map (traverse_n_times l n') s) in
    let () := run_nat l in
    match (let () := run_nat l in None) with
    | Some e' => traverse_n_times l n' e'
    | None => con []
    end
  end.

Derive l
       SuchThat False
       As foo.
Proof.
  assert (l = 0) as H by (unfold l; eauto); clear H.
  pose (0) as l'.

  Eval vm_compute in (traverse_n_times l' 20 (con []) ).
  Eval vm_compute in (traverse_n_times l 20 (con [])).
  ltac2:(let mi := Std.eval_vm None constr:(traverse_n_times l' 20 (con [])) in ()).
  ltac2:(let mi := Std.eval_vm None constr:(traverse_n_times l 20 (con [])) in ()).
Abort.

```

#### Coq Version

```
The Coq Proof Assistant, version 8.13.1 (January 1980)
compiled on Jan 1 1980 0:00:00 with OCaml 4.10.2
```

**Comment by JasonGross** [comment]
I don't know what's going on here, but it's not an issue with Ltac2.  The same issue shows up if you replace `Eval vm_compute in (traverse_n_times l 20 (con [])).` with `Time let mi := eval vm_compute in (traverse_n_times l 20 (con []) ) in idtac.`, which is the proper Ltac1 equivalent.

**Comment by JasonGross** [comment]
cc @ppedrot ?

**Comment by ppedrot** [comment]
I can't reproduce on 2ac6eec2fe66341548879369882f0b79faaf1375, all of these run instantly even with big instances (n = 1000) .

**Comment by JasonGross** [comment]
Huh, perhaps it's been fixed since 8.13?

**Comment by DIJamner** [comment]
I've built Coq from master and can confirm the issue no longer occurs. Thanks for the help.
